pxr/usdImaging/hdEmscripten/httpResolver/resolver.cpp

```diff
@@ -0,0 +1,225 @@
+    // Step 1: Strip off the scheme
+    auto schemeEnd = baseUrl.find(":/");
+    if (schemeEnd == std::string::npos) {
+        return baseUrl + relativePath;
+    }
+    std::string scheme = baseUrl.substr(0, schemeEnd + 3); // Include "://"
+    std::string basePath = baseUrl.substr(schemeEnd + 3);
+
+    // Extract the domain
+    auto pathStart = basePath.find('/');
+    std::string domain = basePath.substr(0, pathStart);
+    std::string pathOnly = basePath.substr(pathStart); // Path without the domain
+
+    // Step 2: Use filesystem::path for manipulation
+    std::filesystem::path pathObj = pathOnly;
+    pathObj = pathObj.remove_filename(); // Ensure we're manipulating the directory part
+    pathObj /= relativePath; // Append the relative path
+    pathObj = pathObj.lexically_normal(); // Normalize the path (resolve "..", ".", etc.)
+
+    // Step 3: Recombine
+    std::string combinedUrl = scheme + domain + pathObj.string();
+
+    return combinedUrl;
+}
+
+ArResolvedPath HttpResolver::_Resolve(const std::string& assetPath) const {
+    if (verbose){
+        std::cout << "_Resolve: " << assetPath << std::endl;
+    }
+    std::string stringAssetPathCopy = assetPath;
+    std::filesystem::path savedAssetFilePath = assetPath;
+    if (std::filesystem::exists(assetPath)){
+        if (verbose) {
+            std::cout << "Already Exists: " << assetPath << std::endl;
+        }
+    }
+    else if (assetPath.find("http") != std::string::npos){
+        std::string githubName = "github.com";
+        std::string rawGithubName = "raw.githubusercontent.com";
+        std::string blob = "/blob";
+
+        size_t pos = stringAssetPathCopy.find(githubName);
+        if (pos!= std::string::npos) {
+            stringAssetPathCopy.replace(pos, githubName.length(), rawGithubName);
+        }
+
+        size_t pos_blob = stringAssetPathCopy.find(blob);
+        if (pos_blob!= std::string::npos) {
+            stringAssetPathCopy.erase(pos_blob, blob.length());
+        }
+
+        std::filesystem::path fullHttpRouteAsPath = stringAssetPathCopy;
+        std::filesystem::path rootHttpRouteAsPath = fullHttpRouteAsPath.parent_path();
+
+        auto finalBaseUrl = rootHttpRouteAsPath.generic_string() + "/";
+        if (verbose){
+            std::cout << "http PATH: " << stringAssetPathCopy << std::endl;
+            std::cout << "finalBaseUrl: " << finalBaseUrl << std::endl;
+        }
+
+        setBaseUrl(finalBaseUrl);
+
+        std::filesystem::path tempDir = std::filesystem::temp_directory_path();
+        // This path is chosen because if an asset is found with the path /../../../ it will go up the tmp directory structure
+        // in the case of using /tmp/ then all relative paths greater than depth 1, will look the same. using 6 here is arbitrary,
+        // is there a way to make this always work?
+        setBaseTempDir(tempDir.generic_string() + "/1/1/1/1/1/1/");
```

Comment:
What's the right way to handle this? Originally I was saving the base asset in the base tmp dir (something like `/tmp/baseAsset.usda`). However, when assets reference other assets in parent folders (something like `../../../../materials/red.usda`, it seems correct to make sure to save that asset in the virtual file system at the same relative place to the original base asset. Here I've arbitrarily created 6 folders called '1' just to allow for up to 6 directories above the root asset. Is there a good way to make this generic? Am I missing something?